src/main/java/io/pravega/connectors/flink/FlinkPravegaWriter.java

```diff
@@ -55,106 +73,464 @@
+    }
+
+    protected String name() {
+        return getRuntimeContext().getTaskNameWithSubtasks();
+    }
+
+
+    // ------------------------------------------------------------------------
+    //  serializer
+    // ------------------------------------------------------------------------
+
+    private static final class FlinkSerializer<T> implements Serializer<T> {
+
+        private final SerializationSchema<T> serializationSchema;
+
+        FlinkSerializer(SerializationSchema<T> serializationSchema) {
+            this.serializationSchema = serializationSchema;
+        }
+
+        @Override
+        public ByteBuffer serialize(T value) {
+            return ByteBuffer.wrap(serializationSchema.serialize(value));
+        }
+
+        @Override
+        public T deserialize(ByteBuffer serializedValue) {
+            throw new IllegalStateException("deserialize() called within a serializer");
+        }
+    }
+
+    // ------------------------------------------------------------------------
+    //  utilities
+    // ------------------------------------------------------------------------
+
+    private static final class TransactionAndCheckpoint<T> {
+
+        private final Transaction<T> transaction;
+        private final long checkpointId;
+
+        TransactionAndCheckpoint(Transaction<T> transaction, long checkpointId) {
+            this.transaction = transaction;
+            this.checkpointId = checkpointId;
+        }
+
+        Transaction<T> transaction() {
+            return transaction;
+        }
+
+        long checkpointId() {
+            return checkpointId;
+        }
+
+        @Override
+        public String toString() {
+            return "(checkpoint: " + checkpointId + ", transaction: " + transaction.getTxnId() + ')';
+        }
+    }
+
+    /*
+     * Interface for exactly once and at least once writerImpl implementation
+     */
+    private interface InternalWriter<T> {
+
+        void initialize() throws Exception;
+
+        void write(T event) throws Exception;
+
+        void close() throws Exception;
+
+        List<UUID> snapshotState(long checkpointId, long checkpointTime) throws Exception;
+
+        void restoreState(List<UUID> list) throws Exception;
+
+        void notifyCheckpointComplete(long checkpointId) throws Exception;
+
+    }
+
+    /*
+     * Exactly-Once Pravega Writer implementation
+     */
+    private class TransactionWriter implements InternalWriter<T> {
+
+        /** The currently running transaction to which we write */
+        private transient Transaction<T> currentTxn;
+
+        /** The transactions that are complete from Flink's view (their checkpoint was triggered),
+         * but not fully committed, because their corresponding checkpoint is not yet confirmed */
+        private transient ArrayDeque<TransactionAndCheckpoint<T>> txnsPendingCommit;
+
+        @Override
+        public void initialize() throws Exception {
```

Comment:
Maybe you should move the logic in `initialize` to the writerImpl constructor.   Then you could mark `txnsPendingCommit` as final.